include/onnxruntime/core/session/onnxruntime_c_api.h

```diff
@@ -696,6 +696,10 @@ typedef void (*OrtCustomJoinThreadFn)(OrtCustomThreadHandle ort_custom_thread_ha
 
 typedef OrtStatus*(ORT_API_CALL* RegisterCustomOpsFn)(OrtSessionOptions* options, const OrtApiBase* api);
 
+typedef void (*RunAsyncCallbackFn)(void*, OrtValue**, size_t, OrtStatusPtr);
```

Comment:
Need to document the params.

Reply:
Done in latest commit.

Reply:
There is an important thing regarding the second parameter of type `OrtValue**`.

Unlike `OrtRun()`, the current design of `OrtRunAsync()` does not allow users to pass in a pre-allocated buffer to accept (number of output) of `OrtValue *`. So this means, in implementation, it's ORT's responsibility to manage the life-cycle of the `OrtValue**` param.

Without the capability to allow users to pass there predefined `OrtValue`s, using pinned buffer or IO binding may not working.

I am not sure if this is an issue. But if it is, one solution is to allow users to pass `OrtValue**` in parameters of `OrtRunAsync()` ( just like `OrtRun()` do), and callback params only contain `user_data` and `status`.

Reply:
There isn't an issue - For RunAsync, the OrtValue** are still pre-allocated with all nullptr(s).

Reply:
Any reason to disallow pre-allocated OrtValues?

Reply:
As per discussion let's keep the interface similar to Run with the addition of a callback + userdata. We can then fill the user provided array of OrtValues and return the same array to the user in the callback. This helps with memory management of the array and the keeps the API consistent with Run allowing users to switch between sync and async with less effort.